From db2dcce3ee28b133315a9294862e570cf608f293 Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Tue, 5 Sep 2017 16:07:19 +0200
Subject: [PATCH 1/1] efi_loader: allow multiple source files for EFI apps

With this patch an EFI application can be built
out of multiple source files.

All object files that are to be included into the EFI
application %.efi must be added to variable %-objs. E.g.

	helloworld-objs = helloworld.o

script/Makefile.lib automatically generates file %_efi.d
containing the dependency definition.

This file is included in the next run of make. From now on
all objects in %-objs are built.

The %_efi.d file should be included in the source
distribution.

After adding a new file to %-objs the first make run will
fail due to the outdated %_efi.d file.

elf2efi is used to create the efi applciation instead of
objcopy.

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
 arch/arm/config.mk              |  2 +-
 arch/x86/config.mk              |  2 +-
 lib/efi_loader/Makefile         |  3 +++
 lib/efi_loader/helloworld_efi.d |  1 +
 scripts/Makefile.lib            | 23 +++++++++++++----------
 5 files changed, 19 insertions(+), 12 deletions(-)
 create mode 100644 lib/efi_loader/helloworld_efi.d

diff --git a/arch/arm/config.mk b/arch/arm/config.mk
index 1a77779db4..d9171b0d8a 100644
--- a/arch/arm/config.mk
+++ b/arch/arm/config.mk
@@ -14,7 +14,7 @@ endif
 endif
 
 CFLAGS_NON_EFI := -fno-pic -ffixed-r9 -ffunction-sections -fdata-sections
-CFLAGS_EFI := -fpic -fshort-wchar
+CFLAGS_EFI := -fno-PIE -fshort-wchar
 
 LDFLAGS_FINAL += --gc-sections
 PLATFORM_RELFLAGS += -ffunction-sections -fdata-sections \
diff --git a/arch/x86/config.mk b/arch/x86/config.mk
index fccb14fd06..d802801dda 100644
--- a/arch/x86/config.mk
+++ b/arch/x86/config.mk
@@ -47,7 +47,7 @@ OBJCOPYFLAGS_EFI := -j .text -j .sdata -j .data -j .dynamic -j .dynsym \
 ifeq ($(IS_32BIT),y)
 CFLAGS_NON_EFI := -mregparm=3
 endif
-CFLAGS_EFI := -fpic -fshort-wchar
+CFLAGS_EFI := -fno-PIE -fshort-wchar
 
 ifeq ($(CONFIG_EFI_STUB_64BIT),)
 CFLAGS_EFI += $(call cc-option, -mno-red-zone)
diff --git a/lib/efi_loader/Makefile b/lib/efi_loader/Makefile
index 5200497230..d6a9635cc8 100644
--- a/lib/efi_loader/Makefile
+++ b/lib/efi_loader/Makefile
@@ -10,6 +10,9 @@
 CFLAGS_helloworld.o := $(CFLAGS_EFI)
 CFLAGS_REMOVE_helloworld.o := $(CFLAGS_NON_EFI)
 
+helloworld-objs = \
+helloworld.o
+
 ifneq ($(CONFIG_CMD_BOOTEFI_HELLO_COMPILE),)
 always += helloworld.efi
 endif
diff --git a/lib/efi_loader/helloworld_efi.d b/lib/efi_loader/helloworld_efi.d
new file mode 100644
index 0000000000..fc70d1e383
--- /dev/null
+++ b/lib/efi_loader/helloworld_efi.d
@@ -0,0 +1 @@
+lib/efi_loader/helloworld_efi.dep: lib/efi_loader/helloworld.o
diff --git a/scripts/Makefile.lib b/scripts/Makefile.lib
index ebc74f8987..f71b1447b6 100644
--- a/scripts/Makefile.lib
+++ b/scripts/Makefile.lib
@@ -363,24 +363,27 @@ cmd_S_efi=					\
 $(obj)/%_efi.S: $(obj)/%.efi
 	$(call cmd,S_efi)
 
-quiet_cmd_efi_objcopy = OBJCOPY $@
-cmd_efi_objcopy = $(OBJCOPY) -j .header -j .text -j .sdata -j .data -j \
-		.dynamic -j .dynsym  -j .rel* -j .rela* -j .reloc \
-		$(if $(EFI_TARGET),$(EFI_TARGET),-O binary) $^ $@
+cmd_efi_objcopy = $(ELF2EFI) --subsystem 10 $^ $@
 
-$(obj)/%.efi: $(obj)/%_efi.so
+$(obj)/%.efi: $(obj)/%_efi.a
 	$(call cmd,efi_objcopy)
 
-quiet_cmd_efi_ld = LD      $@
-cmd_efi_ld = $(LD) -nostdlib -znocombreloc -T $(EFI_LDS_PATH) -shared \
-		-Bsymbolic $^ -o $@
+cmd_efi_ld = $(LD) -q -S -static -T $(EFI_LDS_PATH) -e efi_main \
+		-Bsymbolic $(foreach _s, $($*-objs), $(obj)/$(_s)) \
+		-o $@
 
 EFI_LDS_PATH = $(srctree)/arch/$(ARCH)/lib/$(EFI_LDS)
 
-$(obj)/%_efi.so: $(obj)/%.o arch/$(ARCH)/lib/$(EFI_CRT0) \
-		arch/$(ARCH)/lib/$(EFI_RELOC)
+$(obj)/%_efi.a: $(obj)/%_efi.dep
+	@echo $(obj)/$*_efi.dep: $(foreach _s, $($*-objs), $(obj)/$(_s)) \
+		> $(obj)/$*_efi.d
 	$(call cmd,efi_ld)
 
+$(obj)/%_efi.dep:;
+
+include $(wildcard $(foreach _s, \
+		     $(filter %.efi, $(always)), $(_s:%.efi=$(srctree)/%_efi.d)))
+
 # ACPI
 # ---------------------------------------------------------------------------
 quiet_cmd_acpi_c_asl= ASL     $<
-- 
2.11.0

